kv/store_index.go

```diff
@@ -75,29 +74,97 @@ func (s *IndexStore) DeleteEnt(ctx context.Context, tx Tx, ent Entity) error {
 }
 
 // Find provides a mechanism for looking through the bucket via
-// the set options.
+// the set options. When a prefix is provided, the prefix is used with
+// the index, and not the entity bucket. If you wish to look at the entity
+// bucket and seek, then nest into the EntStore here and do that instead.
 func (s *IndexStore) Find(ctx context.Context, tx Tx, opts FindOpts) error {
 	span, ctx := tracing.StartSpanFromContext(ctx)
 	defer span.Finish()
 
-	if len(opts.Prefix) > 0 {
-		entCaptureFn := opts.CaptureFn
-		opts.CaptureFn = func(key []byte, indexVal interface{}) error {
-			ent, err := s.IndexStore.DecodeToEntFn(key, indexVal)
-			if err != nil {
-				return errors.Wrap(err, "index lookup")
-			}
+	if len(opts.Prefix) == 0 {
+		return s.EntStore.Find(ctx, tx, opts)
+	}
 
-			entVal, err := s.EntStore.FindEnt(ctx, tx, ent)
-			if err != nil {
-				return errors.Wrap(err, "entity lookup")
+	entCaptureFn := opts.CaptureFn
+	kvStream, filterFn := s.indexFilterStream(ctx, tx, opts.FilterEntFn)
+	opts.FilterEntFn = filterFn
+	opts.CaptureFn = func(key []byte, indexVal interface{}) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case fromFilter, ok := <-kvStream:
+			if !ok {
+				return nil
 			}
-			return entCaptureFn(key, entVal)
+			if fromFilter.err != nil {
+				return fromFilter.err
+			}
+			return entCaptureFn(fromFilter.k, fromFilter.v)
+		}
+	}
+	return s.IndexStore.Find(ctx, tx, opts)
+
+}
+
+func (s *IndexStore) indexFilterStream(ctx context.Context, tx Tx, entFilterFn FilterFn) (<-chan struct {
+	k   []byte
+	v   interface{}
+	err error
+}, func([]byte, interface{}) bool) {
+	kvStream := make(chan struct {
+		k   []byte
+		v   interface{}
+		err error
+	}, 1)
+
+	type kve struct {
+		k   []byte
+		v   interface{}
+		err error
+	}
+
+	send := func(key []byte, v interface{}, err error) bool {
+		select {
+		case <-ctx.Done():
+		case kvStream <- kve{
+			k:   key,
+			v:   v,
+			err: err,
+		}:
 		}
-		return s.IndexStore.Find(ctx, tx, opts)
+		return true
 	}
 
-	return s.EntStore.Find(ctx, tx, opts)
+	return kvStream, func(key []byte, indexVal interface{}) (isValid bool) {
+		defer func() {
+			if !isValid {
+				close(kvStream)
+			}
+		}()
+		ent, err := s.IndexStore.DecodeToEntFn(key, indexVal)
+		if err != nil {
+			return send(nil, nil, err)
+		}
+
+		entVal, err := s.EntStore.FindEnt(ctx, tx, ent)
```

Comment:
this should always be the id, right?

Reply:
I'm a little lost. Which part is suppose to be the `id`? The entity that you see really just the ID. The abstraction works against the `Entity` type though. What this is doing is a conversion between ID to an entity. This one ends up being that the index val is an ID, but that may not always be the case.